Paper:  ITK-Wasm Universal spatial analysis and visualization

thewtex opened the pull request:
_Editor_: <!--editor--> Sanhita Joshi @sanhitamj  <!--end-editor-->

_Reviewers_: <!--reviewers-list-->
 - Andrew James @amjames 
 - Pragnya Modi @pragyamodi
 <!--end-reviewers-list-->

thewtex committed:
Template + id update

thewtex committed:
Update myst.yml

And abstract.

thewtex committed:
Introduction

thewtex committed:
Initial draft

thewtex committed:
Text tweaks

thewtex committed:
Example Visible Male data

thewtex committed:
Citation updates

thewtex committed:
Test plugin

thewtex committed:
Build errors

thewtex committed:
environment.yml, package.json pnpm-lock.yaml

thewtex committed:
Heading improvements

fwkoch commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+It allowed developers to write code in languages like C and C++, compile it to asm.js, and run it in the browser with near-native performance.
+Asm.js achieved this by using a statically-typed subset of JavaScript that enabled optimizations by the JavaScript engine.
+
+:::{unsplash} melody
```
I believe unsplash recently disabled the api used by this plugin (and their new one now requires an auth token...). That means the next time your paper is built, this will not work.

Can you just replace this with a static image? (Or just eliminate it entirely?)

amjames replied:
In the current preview a picture of a saxophone on a padded bench is rendered. I am unsure how this relates to the content of the section. If it should stay it would be appropriate to generate a figure caption and refer to it in the relevant text.

amjames replied:
In the current preview a picture of a saxophone on a padded bench is rendered. I am unsure how this relates to the content of the section. If it should stay please generate a figure caption and refer to it in the relevant text.

thewtex replied:
This was experiment to test MyST renderers on SciPy proceedings. Removed now :-)

Thanks to support from @stevejpurves , we aim to demonstrate SciPy's advancements in reproducible, interactive publishing  via https://github.com/curvenote/itk-plugins :sparkler: :man_astronaut:

amjames commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+The Insight Toolkit (ITK) is an open-source, cross-platform library that provides developers with an extensive suite of software tools based on a proven, spatially-oriented architecture for processing scientific data in two, three, or more dimensions [@doi:10.3389/fninf.2014.00013;].
+ITK includes fundamental numerical libraries, such as Eigen.
+ITK's C++ template-based architecture inherently helps keep Wasm modules small while enabling the compiler to add extensive performance optimizations.
+The _itk-wasm_ GitHub repository is also an ITK Remote Module, `WebAssemblyInterface`, that implements Wasm-interface specific functionality.
```
This could use some additional information to explain the significance of the GH repo being a remote module.

thewtex replied:
More added in https://github.com/scipy-conference/scipy_proceedings/pull/927/commits/bc8b0e9c022bd5567974be459dddeb6b84c4d0c3

amjames commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+A notable application of ITK-Wasm is the generation of multiscale OME-Zarr images, a cloud-optimized bioimaging format with broad international adoption [@doi:10.1007/s00418-023-02209-1; @doi:10.1038/s41592-021-01326-w].
+To generate OME-Zarr's multiscale representation of multidimensional bioimages, anti-aliasing filters must be applied.
+
+In the NGFF-Zarr package, ITK-Wasm anti-aliasing filters efficiently produces OME-Zarr images suitable for Pyodide, JupyterLite, and traditional CPython environments [@doi:10.5281/zenodo.8092821].
```
```suggestion
In the NGFF-Zarr package, ITK-Wasm anti-aliasing filters efficiently produce OME-Zarr images suitable for Pyodide, JupyterLite, and traditional CPython environments [@doi:10.5281/zenodo.8092821].
```

amjames commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+
+In the NGFF-Zarr package, ITK-Wasm anti-aliasing filters efficiently produces OME-Zarr images suitable for Pyodide, JupyterLite, and traditional CPython environments [@doi:10.5281/zenodo.8092821].
+Since the interface type's Python representation are Python data classes comprised of standand Python data types and NumPy arrays, they are trivially and efficiently serialized for parallel computing with Dask.
+Furthermore, a cuCIM accelerator package exemplifies ITK-Wasm's compatibility with GPU acceleration. Its utility extends to desktop applications like 3D Slicer, illustrating its versatility and broad applicability in the scientific computing ecosystem.
```
This paragraph begins with a description of the `NGFF-Zarr` package, but ends with the mention of a different `cuCIM` package. It is not clear to me if the later is part of the former, or distinct from it. If they are distinct it should get its own reference, and a separate paragraph highlighting it's relevant features would be great.

thewtex replied:
👍 split into separate paragraphs.

amjames commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+2. **WebAssembly Interface Types**: Standardizes the way Wasm modules interact with each other and with host environments, simplifying the integration process.
+3. **Component Model**: An emerging standard that aims to improve modularity and reuse of Wasm components, further enhancing interoperability .
+
+ITK-Wasm provides a robust framework for scientific computing that leverages WebAssembly's strengths. The framework bridges the gap between web-based and native applications, enabling high-performance, cross-platform scientific analysis. By integrating the principals of the WebAssembly Component Model, ITK-Wasm enhances interoperability and sustainability, allowing scientific Python to thrive in  ...
```
At this point in the paper (as someone who is not very familiar with ITK) it is not clear to me how ITK-Wasm "enhances interoperability and sustainability, allowing scientific Python to thrive in a multi-language ecosystem". 

- Is/are the package(s) mentioned in the Results section multi-language, or are consumers of those outputs typically found in a different language ecosystem than the one used to generate them? 

- Are other solutions for generating those results prone to reproducibility issues, and how does the ITK-WASM based solution address that? 

It is also mentioned here that the framework enables high-performance in scientific analysis, but there is not much said about that in the results section. Is this quantifiable in any way? Is the achievement notable because cross-platform/multi-language support is achieved without sacrificing performance?

thewtex replied:
I fleshed out the example in the Results section, detailing a motivating use case and how the associated code and packages look.

Regarding reproducibility, I added another paragraph to expand on the connection:

> 
Our approach excels in sustainability and composability, thanks to small, self-contained, and idiomatic packages that are platform-agnostic and have minimal dependencies. This design enables outstanding computational reproducibility.

I will look into the possibility of performance quantification.

thewtex replied:
> It is also mentioned here that the framework enables high-performance in scientific analysis, but there is not much said about that in the results section. Is this quantifiable in any way? Is the achievement notable because cross-platform/multi-language support is achieved without sacrificing performance?

I added two new figures and captions that discuss the issues of performance, accessibility, and sustainability.

amjames commented on papers/matt_mccormick/main.md:
```diff
@@ -0,0 +1,352 @@
+Build scripts are provided to build TypeScript to JavaScript and also generate a demo app with an HTML interface.
+JavaScript bindings load Zstandard-compressed versions the Wasm modules on-demand in a web worker to support progressive and performant execution.
+
+Python packages are generate for both system execution and web brower-execution. In the browser, Pyodide-compatible packages provide client-side web app scripting in Python, including via PyScript, and sustainable, scalable Jupyter deployments via JupyterLite.
```
```suggestion
Python packages are generated for both system execution and web brower-execution. In the browser, Pyodide-compatible packages provide client-side web app scripting in Python, including via PyScript, and sustainable, scalable Jupyter deployments via JupyterLite.
```

Is this section describing bindings generated for the core `itk-wasm` package, bindings generated by the `itk-wasm` build system for downstream packages based on it, or both?

thewtex replied:
downstream packages -> I added `for new modules`.

amjames commented on papers/matt_mccormick/main.md:
````diff
@@ -0,0 +1,352 @@
+npx create-itk-wasm
+```
+
+The `create-itk-wasm` tool, which can be used interactively or programmatically, will generate C++ code with the desired ITK-Wasm CLI11 interfaces, other support configuration files such as CMake build configuration scripts and a conda _environment.yml_ file, and a PNPM _package.json_ file that will drive Wasm module builds, language binding generation, and testing. The PNPM workspace internally  ...
````
> ..., will generate C++ code with the desired ITK-Wasm CLI11 interfaces

Is the word "desired" the best choice here? If the user can chose which interfaces are needed, then how is that controlled? If the interfaces are the ones needed for proper integration then perhaps "required" would make this sentence more clear.

thewtex replied:
:+1: changed to "required"

amjames reviewed (commented):
I am familiar with WASM, but not ITK. 

My initial impressions as a partial outsider are:
- That this paper does a great job introducing the foundational technology upon which ITK-Wasm is developed. 
- The methods section could benefit from some examples of how to use various components which are outlined. 
- The results/discussion are a bit weak with respect to supporting the key conclusions of the paper.

Minor fixes and comments are inline, summary of larger suggestions  by section below. 

## Methods
This section could benefit from some concrete examples, or perhaps a single representative example of the end-to-end process split across the description of the various core components. For example if the paper discussed the different components against the backdrop of building a simple workflow that say reads, rotates, and writes an image the reader would have a much better grasp of how these parts fit together, as well as a jumping off point for adapting to their own particular needs. 

In its current state this section restates, in many places verbatim, sections of the documentation without providing much additional context or actionable instruction. 

## Results/Discussion
The opening and closing paragraphs imply the paper is a discussion of how ITK-Wasm provides improvements to a few key areas in scientific computing:  

- Interoperability
- Performance
- Sustainability

There is a very strong example of sustainability provided in the introduction referring to the interactive figures application build with the *predecessor* to ITK-Wasm. In the remaining sections not much is said to support the other claims until they are reiterated in the conclusion.

This submission could benefit from some expansion of the results and discussion sections to call out examples of workloads where implementation based on ITK-Wasm made improvements in these three areas.

thewtex committed:
Remove unsplash renderer

thewtex commented:
@amjames thank you for the review! I agree with your comments and will follow-up accordingly.

thewtex committed:
Note RapidJSON to Glaze

thewtex committed:
Add ITK Remote Module relevance

thewtex committed:
Update papers/matt_mccormick/main.md

Co-authored-by: Andrew James <Andrew.m.james2@gmail.com>

thewtex committed:
environment.yml -> pixi.toml

thewtex committed:
Add OME-Zarr overview figure

sanhitamj commented:
Hi @thewtex, 2nd Sept is the final deadline to make any edits to the manuscript.

thewtex committed:
Add vm-head-frozenct-ome-zarr figure

Add Paul Elliott as co-author, helped with the figure, worked on
ITK-Wasm.

thewtex commented:
@stevejpurves we could forgo the renderer plugin until it is more mature and just use a figure and iframe directive as you suggested. But how do I use the `placeholder` + `figure` + `iframe` so we do not get the placeholder rendered in HTML?

I currently get:

![image](https://github.com/user-attachments/assets/1893b4f4-a5b7-410f-98fb-a1d53d221fed)

with:

::::{figure}
:label: fig:vm_head_frozenct_ome_zarr
:placeholder: figures/vm-head-frozenct-ome-zarr-snapshot.png

Visible Male frozen head computed tomography (CT) OME-Zarr volume, generated and visualized with ITK-Wasm. There are three resolution scales, which can be selected with the _Image Scale_ buttons. Reduced resolutions are smoothed with a gaussian filter to avoid aliasing artifacts.

:::{iframe} https://scipy-2024-itk-wasm-vm-head-frozenct-ome-zarr.netlify.app/
:width: 100%
:::

::::

thewtex committed:
Antialias filter description

thewtex committed:
aliasing figure inputs

thewtex committed:
Start anti-alias figure

thewtex committed:
Add Paul's ORCID

thewtex committed:
Set normalizedOpacityPoints

thewtex committed:
Aliasing comparison figure

thewtex committed:
Address curvenote lint errors

thewtex committed:
C++ pipeline definition

thewtex committed:
Remove netlify auto-generated .gitignore content

thewtex committed:
Missing VSM VisibleHumanMale DOI override

thewtex committed:
Command line invocation

thewtex committed:
Typescript bindings

thewtex committed:
downsample Python package description

thewtex commented:
@stevejpurves @rowanc1 how do we make figure placeholder images (PDF generation) work with iframes?

thewtex committed:
review updates

thewtex committed:
Clarifications re: review

rowanc1 commented:
If you can include the image you would like and a comment above the iframe in the markdown, we will take care of it at publication time.

sanhitamj commented:
Hi @thewtex thanks for the submission and the subsequent edits to the paper. Tomorrow, 2nd Sept is the last day for author edits.

Hi @amjames thank you for helping with reviewing the paper. Will you please share your final decision about the paper by 9th Sept?

thewtex committed:
benchmark figures

thewtex committed:
figure tweaks

thewtex committed:
Comment out figure placeholders

Per [discussion with
@rowanc1](https://github.com/scipy-conference/scipy_proceedings/pull/927#issuecomment-2322775234), these will be included in the PDF at
publication time.

thewtex committed:
narrative tweaks

thewtex commented:
> If you can include the image you would like and a comment above the iframe in the markdown, we will take care of it at publication time.

@rowanc1 thanks! I commented out the placeholder images in https://github.com/scipy-conference/scipy_proceedings/pull/927/commits/044dd4a300c78728c6421d1fc780336045d827eb

Has the SciPy Proceeding MyST theme been made available? I think this would follow the SciPy Conference's FLOSS values. And it would enable folks like myself to look into issues like this when encountered.

> Hi @thewtex thanks for the submission and the subsequent edits to the paper. Tomorrow, 2nd Sept is the last day for author edits.

@sanhitamj thanks for the guidance and help editing! I have pushed all edits.


@amjames thanks again for the review. I tried to address all of your comments.

amjames reviewed (approved):
All of my concerns have been adequately addressed. Thanks!
